storage/grpc_client.go

```diff
@@ -840,6 +841,100 @@ func (c *grpcStorageClient) DeleteHMACKey(ctx context.Context, desc *hmacKeyDesc
 	return errMethodNotSupported
 }
 
+// Notification methods.
+
+func (c *grpcStorageClient) ListNotifications(ctx context.Context, bucket string, opts ...storageOption) (n map[string]*Notification, err error) {
+	ctx = trace.StartSpan(ctx, "cloud.google.com/go/storage.grpcStorageClient.ListNotifications")
+	defer func() { trace.EndSpan(ctx, err) }()
+
+	s := callSettings(c.settings, opts...)
+	if s.userProject != "" {
+		ctx = setUserProjectMetadata(ctx, s.userProject)
+	}
+	req := &storagepb.ListNotificationsRequest{
+		Parent: bucketResourceName(globalProjectAlias, bucket),
+	}
+	var notifications []*storagepb.Notification
+	err = run(ctx, func() error {
+		gitr := c.raw.ListNotifications(ctx, req, s.gax...)
+		for {
+			items, nextPageToken, err := gitr.InternalFetch(int(req.PageSize), req.PageToken)
```

Comment:
I was initially confused about how the page token was being preserved between loop iterations but looks like `InternalFetch` modifies the request to handle this. Might be worth adding a comment about this since it does seem confusing!

Also @noahdietz can you just verify that this will preserve the page token if there is a failure from internalFetch mid-way through? Otherwise we might end up with duplicate results.

Reply:
>  looks like InternalFetch modifies the request to handle this

I'm not sure what you mean by this. `InternalFetch` will return the nextPageToken, but not modify the next request to use it. That's how I [read its code](https://github.com/googleapis/google-cloud-go/blob/main/storage/internal/apiv2/storage_client.go#L704-L725) at least. I think we need to set `req.PageToken = nextPageToken` right after line 869 before the next iteration.

Reply:
Ahh I was not preserving the nextPageToken between iterations >"< updated. Thank you both for the clarification!

Reply:
Hah @noahdietz that's what I thought originally looking through the code and I talked myself out of it for some reason. Yup that sounds like the right answer.